### ISSUE TITLE
Rewrite name resolution

### ISSUE THREAD

**Issue by xeno-by** [twitter/rsc#229]
At the moment, name resolution is deliberately incorrect. The current goal is to get as far as possible with an implementation that is as simple as possible. If the implementation becomes too involved / too big, I'll scrap it and rewrite it from scratch.

**Comment by xeno-by** [comment]
Given that our current approximation of name resolution has become somewhat inaccurate as we increased the scope of the case study project, I bit the bullet and rewrote it. 

Unfortunately, I discovered that the SLS is incorrect - oh well - see the linked commits above for details. Luckily, reverse-engineering Scalac took only a few hours, and the end result is surprisingly compact. It has undefined behavior for invalid code (e.g. when ambiguities or visibility issues would lead to failed name resolutions in Scalac), but we don't care about error reporting yet, since under multi-phase compilation code is always compiled by Scalac anyway.
